extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java

```diff
@@ -95,8 +115,190 @@ public void handle(RoutingContext routingContext) {
+                            String source = (String) invoker.getBindingContext().get(RESPONSE_SOURCE);
+                            outputCloudEvent = CloudEventBuilder.create()
+                                    .setSpecVersion("1.0")
+                                    .setId(getResponseId())
+                                    .setType(type)
+                                    .setSource(URI.create(source))
+                                    .setDataContentType(
+                                            byte[].class.equals(innerOutputType) ? "application/octet-stream"
+                                                    : "application/json")
+                                    .setExtensions(Collections.emptyMap())
+                                    .setData(output)
+                                    .build();
+                        } else {
+                            // user is explicitly returning CloudEvent
+                            outputCloudEvent = (CloudEvent<?>) output;
+                        }
+
+                        if (binaryCE) {
+                            httpResponse.putHeader("ce-id", outputCloudEvent.id());
+                            httpResponse.putHeader("ce-specversion", outputCloudEvent.specVersion());
+                            httpResponse.putHeader("ce-source", outputCloudEvent.source().toString());
+                            httpResponse.putHeader("ce-type", outputCloudEvent.type());
+
+                            if (outputCloudEvent.time() != null) {
+                                httpResponse.putHeader("ce-time", outputCloudEvent.time().toString());
+                            }
+
+                            if (outputCloudEvent.subject() != null) {
+                                httpResponse.putHeader("ce-subject", outputCloudEvent.subject());
+                            }
+
+                            outputCloudEvent.extensions()
+                                    .forEachRemaining(e -> httpResponse.putHeader("ce-" + e.getKey(), e.getValue()));
+
+                            String dataContentType = outputCloudEvent.dataContentType();
+                            if (dataContentType != null) {
+                                httpResponse.putHeader("Content-Type", dataContentType);
+
+                            }
+                            if (dataContentType != null && dataContentType.startsWith("application/json")) {
+                                httpResponse.end(Buffer.buffer(writer.writeValueAsBytes(outputCloudEvent.data())));
+                            } else if (byte[].class.equals(innerOutputType)) {
+                                httpResponse.end(Buffer.buffer((byte[]) outputCloudEvent.data()));
+                            } else {
+                                log.errorf("don't know how to write ce to output (dataContentType: %s, javaType: %s)",
+                                        dataContentType, innerOutputType);
```

Comment:
shout it be caught at build time .. I am just wondering if we could make it more user friendly as only json and binary types are supported so waiting until function invocation fail on it sounds bit too late.. though not sure what options we have... maybe data content type could be an enum ...just thinking at loud.

Reply:
Data content type should not be an enum.  A user could do this pseudocode:

byte[] bytes = protobuf.marshal(object);
builder.setDataType("application/protobuf")
builder.setData(bytes).

Maybe the Builder.setDataType() method should be removed and replaced with something like this?

Builder {
   public Builder binaryData(byte[] data, String dataType);
}

Using Builder.data(Object obj) assumes that default marshalling mechanisms are used.  Usually json, but we may offer plugging in a default marshaller.

Reply:
so we would have these two on the builder

````
public Builder binaryData(byte[] data, String dataType);
public Builder data(Object data);
````

so the first one won't do any marshalling of the data and set the data content type of the event to the given value regardless of what that is
the second one will do json marshalling and set `application/json` as data content type.
Did I got it right?

Reply:
Correct.  What is your opinion though?  With my suggestion, it prevents the user from thinking that we support different content types.

Reply:
I actually like it as it allows to use any content type but the marshalling is on caller - function itself. so +1 from me on this.

Reply:
If I add `binaryData(byte[] data, String dataType);` into current builder, one could call `binaryData` on non-binary builder. Maybe we should have two builder classes. One for object data and the other one for binary?

Reply:
Or maybe I could make builder non-generic and only `build()` methods would be generic, there would be three `build()` methods: a) `build()` no-args for no data, b), `build(T obj)` and `build(byte[], String)`?

Reply:
Does it make sense to send cloud event without data? Just thinking if no arg build would be useful?

mother then that I like idea with separate build methods for data of the event

Reply:
`As defined by the term Data, CloudEvents MAY include domain-specific information about the occurrence. ` I believe data are optional, also data may be passed indirectly using extension: https://github.com/cloudevents/spec/blob/v1.0/extensions/dataref.md (useful probably in case when data are too big).****

Reply:
Fair enough, data is optional so having no arg makes sense then, thanks

Reply:
@patriot1burke is it ok to have multiple build methods with parameters on builder?

Reply:
I am thinking of having something like:
```java
public CloudEvent<byte[]> build(byte[] data, String dataContentType) {
    return new SimpleCloudEvent(specVersion, id, type, source, dataContentType, data, /*has data*/ true, subject, time, extensions);
}

public <T> CloudEvent<T> build(T data) {
    return new SimpleCloudEvent(specVersion, id, type, source, "application/json", data, true, subject, time, extensions);
}

public CloudEvent<Void> build() {
    return new SimpleCloudEvent(specVersion, id, type, source, null, null, false, subject, time, extensions);
}
```
What bothers me is that `public <T> CloudEvent<T> build(T data)` still can be called with `byte[]`. I wish I could do something like `C++` template specialization to prevent that, but in Java it's probably not possible.

Reply:
Maybe calling `public <T> CloudEvent<T> build(T data)` with `T=byte[]` could be sometimes valid, but I think it's not what user would like to do in most cases.

Reply:
I like the above signatures so long as there is no setter for dataContentType.  As I mentioned before there should be defaults for type and source and id if not specified and be consistent with the current defaults described in the Funqy Knative Events docs.

Reply:
+1 on these signatures, clear and simple to use in my opinion